src/mercury-agent/mercury/agent/procedures/hp_firmware.py

```diff
@@ -0,0 +1,259 @@
+    """
+    hpsum_log_path = '/var/hp/log/localhost/hpsum_log.txt'
+    exit_code = None
+    with open(hpsum_log_path) as f:
+        contents = f.readlines()
+
+    for line in contents:
+        if 'Exit Code' in line:
+            exit_code = line.split(':')[1]
+
+    status = get_return_code_status(int(exit_code))
+    log.info(status.get('message'))
+    if status.get('error_state'):
+        raise HPFirmwareException('Flashing failed with exit code {0}'.format(
+            exit_code))
+    return status.get('reboot_flag')
+
+
+@capability('hp_apply_bios_settings',
+            description='Apply bios settings found in given file',
+            kwargs_name=['url'], serial=False,
+            dependency_callback=is_hp, timeout=60)
+def hp_apply_bios_settings(url=None):
+    """
+    Apply BIOS settings found at the given URL
+
+    :param url: Full URL to the BIOS file
+    """
+    # TODO: Implement this
+    # TODO: Alongside the conrepfile, some blobs have to be installed to bring
+    # certain components up to a recommended version. Who determines which blobs
+    # to execute? Are they provided or does this method
+    # keep track of those based on the model 
+
+    # DL380Gen9 = HPServer(name="ProLiant DL380 Gen9",
+    #                      cfg='DL380Gen9_conrep_RAX.dat',
+    #                      blobs=['BCM5719_2.13.5_RHEL6.scexe',
+    #                             'HPSA_1.18_RHEL6.scexe',
+    #                             'iLO4_2.02_RHEL6.scexe'])
+
+
+@capability('hp_update_firmware',
+            description='Installs updates from provided packages',
+            kwarg_names=['url', 'dry_run'], serial=False,
+            dependency_callback=is_hp, timeout=3600)
+def hp_update_firmware(url=None, dry_run=False):
+    """
+    Apply HP firmware updates
+
+    :param url: Full URL to the package containing firmware files
+    :param dry_run: If firmware updates should be applied when found
+    """
+    if not url:
+        return
+
+    try:
+        r = requests.get(url, stream=True, verify=False)
+    except requests.RequestException as err:
+        log.error('There was an error in retrieving the '
+                  'firmware package: {}'.format(err))
+        return
+
+    download_path = '/tmp/hpfirmware.tar.gz'
+    with open(download_path, 'wb') as f:
+        for chunk in r.iter_content(1024**2):
+            if not chunk:
+                break
+            f.write(chunk)
+
+    if not _extract(download_path, firmware_path):
+        return
+
+    updates_available = find_available_updates()
+    if dry_run:
+        if updates_available:
+            log.info('Updates were available but were not installed')
+        return
+
+    # Return code could be a false positive, check log file output
+    install_updates(updates_available)
+    reboot_required = parse_log_output_for_errors()
+    # TODO: Implement a reboot capability?
+    if reboot_required:
+        log.debug('A reboot is required')
```

Comment:
Consider building a return structure. Mercury has zero orchestration, so this should simply return something like:
```
{'result': {
  'return_code': 1,
  'success': True,
  'reboot_required': True
}}
```
Whatever called it can make the decision to reboot the server or not.